hydra-node/src/Hydra/Ledger/Cardano.hs

```diff
@@ -95,712 +82,510 @@ cardanoLedger =
-  generatorEnv = (genEnv Proxy){geConstants = noPPUpdatesTransactions}
+--
+-- Utxo
+--
 
-genSequenceOfValidTransactions :: Utxo CardanoTx -> Gen [CardanoTx]
-genSequenceOfValidTransactions initialUtxo
-  | initialUtxo == mempty = pure []
-  | otherwise = do
-    n <- getSize
-    numTxs <- choose (1, n)
-    genFixedSizeSequenceOfValidTransactions numTxs initialUtxo
+type Utxo = Utxo' LedgerEra (Ledger.Alonzo.TxOut LedgerEra)
 
-genFixedSizeSequenceOfValidTransactions :: Int -> Utxo CardanoTx -> Gen [CardanoTx]
-genFixedSizeSequenceOfValidTransactions numTxs initialUtxo
-  | initialUtxo == mempty = pure []
-  | otherwise = do
-    reverse . snd <$> foldM newTx (initialUtxo, []) [1 .. numTxs]
- where
-  newTx (utxos, acc) _ = do
-    tx <- genCardanoTx utxos
-    case applyTransactions cardanoLedger utxos [tx] of
-      Left err -> error $ show err
-      Right newUtxos -> pure (newUtxos, tx : acc)
+-- | Newtype with phantom types mostly required to work around the poor interface
+-- of 'Ledger.UTXO'and provide 'Monoid' and 'Foldable' instances to make utxo
+-- manipulation bareable.
+newtype Utxo' era out = Utxo
+  { unUtxo :: Map (Ledger.TxIn (Ledger.Crypto era)) out
+  }
+  deriving newtype (Eq, Show)
 
-type CardanoEra = MaryEra StandardCrypto
+instance ToCBOR Utxo where
+  toCBOR = toCBOR . coerce @_ @(Ledger.UTxO LedgerEra)
+  encodedSizeExpr sz _ = encodedSizeExpr sz (Proxy @(Ledger.UTxO LedgerEra))
 
-type CardanoTxBody = Cardano.TxBody CardanoEra
+instance FromCBOR Utxo where
+  fromCBOR = coerce @(Ledger.UTxO LedgerEra) <$> fromCBOR
+  label _ = label (Proxy @(Ledger.UTxO LedgerEra))
 
-type CardanoTxWitnesses = Cardano.WitnessSet CardanoEra
+instance Functor (Utxo' era) where
+  fmap fn (Utxo u) = Utxo (fmap fn u)
 
-type CardanoAuxiliaryData = StrictMaybe (AuxiliaryData CardanoEra)
+instance Foldable (Utxo' era) where
+  foldMap fn = foldMap fn . unUtxo
+  foldr fn zero = foldr fn zero . unUtxo
 
-type CardanoAddress = Cardano.Addr StandardCrypto
+instance Semigroup Utxo where
+  Utxo uL <> Utxo uR = Utxo (uL <> uR)
 
-type CardanoKeyPair = Cardano.KeyPair 'Cardano.Payment StandardCrypto
+instance Monoid Utxo where
+  mempty = Utxo mempty
 
-type TxIn = Cardano.TxIn StandardCrypto
+instance ToJSON Utxo where
+  toJSON = toJSON . unUtxo
 
-type TxOut = Cardano.TxOut CardanoEra
+instance FromJSON Utxo where
+  parseJSON = fmap Utxo . parseJSON
 
-signWith ::
-  Cardano.TxId StandardCrypto ->
-  Cardano.KeyPair 'Cardano.Payment StandardCrypto ->
-  Cardano.WitVKey 'Cardano.Witness StandardCrypto
-signWith (Cardano.TxId safeHash) credentials =
-  Cardano.WitVKey
-    (asWitness $ Cardano.vKey credentials)
-    (signedDSIGN @StandardCrypto (Cardano.sKey credentials) (extractHash safeHash))
+instance Arbitrary Utxo where
+  -- TODO: shrinker!
+  arbitrary = genUtxo
 
---
--- Balance
---
+prettyUtxo :: (TxIn, TxOut ctx era) -> Text
+prettyUtxo (k, TxOut _ (txOutValueToValue -> v) _) =
+  T.drop 54 (renderTxIn k) <> " ↦ " <> prettyValue v
```

Comment:
Must: Can you add a comment why `54` is fine?

Reply:
Yes, though the function pre-existed this PR :sweat_smile: